[Core][Model] PrithviMAE Enablement on vLLM v1 engine (superseded by PR 20577)

christian-pinto opened the pull request:
## Purpose
This PR enables the Prithvi MAE Geospatial model on the vLLM V1 engine. This is a result of pooling/embedding models being enabled in V1.

The main changes in this PR are:
- Adding support for multi-modal models that require the raw multi-modal data in input and not the multi_modal embeddings.
- Improving support for models that use `skip_tokenizer_init=True`
- Improving support for attention-less models by creating a `DummyKVCacheManager` class
- Minor changes to the PrithviMAE model and the example usage script.

## Test Plan
I have added a test script that is only testing the with vLLM runner as this model is not on HF. Therefore it would only check if the inference takes place correctly for the model.

## Test Result
The test is not succeeding at the moment because the pytest process hangs indefinitely while initializing the vlLM engine. The test hangs indefinitely on `wait_for_engine_startup`. The same code tested outside of pytest works just fine. Any help on getting this test running would be much appreciated.

## Documentation Update
Updated examples file.

@DarkLight1337 @maxdebayser @njhill

christian-pinto committed:
Support for attention free models in V1

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Better support for skip_tokenizer_init=True

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Support passing raw multimodal data to model

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Support for attention free models in V1

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Last few changes after rebasing to latest branch version

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
latest changes to align with the original branch

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Latest changes to aadpt to upstream master

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Some reformatting to make the pre-commit hooks succeed

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Few more changes to solve some other pre-commit hooks failures

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -64,7 +65,130 @@ def new_empty(self) -> "KVCacheBlocks":
 
 
-class KVCacheManager:
+class KVCacheManagerInterface(ABC):
```
Not sure whether this conforms with V1 design. cc @maxdebayser @heheda12345 @WoosukKwon

christian-pinto replied:
I have used the same interface template used for the Scheduler in v1. If another way of defining an interface if the preferred one, please let me know.

maxdebayser replied:
BTW, is there anyone working on pure SSM model support in V1?

DarkLight1337 replied:
> @DarkLight1337 , by V1 design do you mean "the right way of defining interfaces" or whether the KVCacheManager should be pluggable or not?

Yeah, I'm not sure which parts of KVCacheManager should be customizable.

> BTW, is there anyone working on pure SSM model support in V1?

#19327

christian-pinto replied:
I went down the road of "fully overloading" the KVCache manager because it is heavily used in the scheduler. At this stage I wanted to avoid messing too much with the `schdule()` function.

heheda12345 replied:
We may have two options:
1. treat the model as if it has 0 kv_cache_group and reuse the existing logic
2. overload KVCacheManager as in this PR

I can dive into more details about the embedding model workflow and we can talk more about it then.

heheda12345 replied:
Want minor changes do you need? I feel it would be quite clean to treat it as 0 groups but maybe I've missed something.

christian-pinto commented on vllm/config.py:
```diff
@@ -614,6 +614,10 @@ def __post_init__(self) -> None:
         self.served_model_name = get_served_model_name(self.model,
                                                        self.served_model_name)
         self.multimodal_config = self._init_multimodal_config()
+        self.is_pooling_model = self.registry.is_pooling_model(
+            self.architectures)
```
sure, no problem

christian-pinto committed:
Some style changes

- Improved formatting around
- made is_pooling_model a @property in ModelConfig

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

maxdebayser commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -547,6 +554,55 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
+                                                      scheduler_output,
+                                                      num_reqs)
+
+    def _maybe_compute_attn_prefix(
```
What is the purpose of this function? It looks like an extension point for a subclass but it's not called anywhere.

maxdebayser commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -547,6 +554,55 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
+    ) -> list[int]:
+        return [0] * len(self.kv_cache_config.kv_cache_groups)
+
+    def _maybe_prepare_additional_inputs(self,
```
Same as above.

christian-pinto replied:
Same as above 😄

maxdebayser reviewed (commented):
This PR introduces some changes that could be useful for future changes such as the implementation of pure SSM models and encoder embedding models. Currently the design of multi-modal inputs as embeddings seems pretty restrictive to me, but I don't know enough about all the design considerations that went into it. So to my superficial level of knowledge about the multi modal models, the support for raw inputs seems pretty useful.

aarnphm commented on vllm/v1/structured_output/__init__.py:
```diff
@@ -46,11 +46,12 @@ def __init__(self, vllm_config: VllmConfig):
-            scheduler_config=self.vllm_config.scheduler_config,
-            lora_config=self.vllm_config.lora_config,
-        ).get_lora_tokenizer(None)
+        self.tokenizer = None if vllm_config.model_config.skip_tokenizer_init \
```
How is this related to this PR?

btw I we can't skip tokenizer init here given that tokenizer is required for all of the structured outputs backend. therefore, this shouldn't be changed.

christian-pinto replied:
I see, this was one quick and dirty change that I made and forgot to sanitize. 

Since the tokenizer is required for structured output generation the output manager should not be initialized at all when we skip tokenizer init. Also, all the call to the structured output manager should be guarded against it.

Let me give this a go and you can have a look at it.

christian-pinto replied:
@aarnphm I have reworked supporting skip_tokenizer_init so that it takes care of the structured output manager. Please have a look at commit  https://github.com/vllm-project/vllm/pull/20072/commits/30ed44e217bf502df7f59c6ae8ca9f2dc8c64914

aarnphm reviewed (commented):
qq: does this work with transformers implementation?

christian-pinto commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -547,6 +554,55 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
         if batch_changed or batch_reordered:
             self.input_batch.refresh_sampling_metadata()
 
+    def _add_multimodal_inputs_to_model_args(
```
yes, makes sense. Thanks

christian-pinto committed:
Simple code refactoring
- Remove unused functions
- merged functions not called anywhere else

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
@aarnphm I have reworked supporting skip_tokenizer_init so that it takes care of the structured output manager. Please have a look at commit  https://github.com/vllm-project/vllm/pull/20072/commits/30ed44e217bf502df7f59c6ae8ca9f2dc8c64914

aarnphm commented:
> > qq: does this work with transformers implementation?

Ah i wad referring to `--model-impl transformers` but it doesn't seem to work. All good though

aarnphm reviewed (commented):
For models that doesn't support structured outputs, I think we should perform the gate at before server startup time. 

For structured outputs overhead, I don't think `skip_tokenizer_init` is relevant here given that the manager initialize the tokenizer itself.

christian-pinto commented on vllm/v1/engine/core.py:
```diff
@@ -86,7 +86,11 @@ def __init__(self,
                             args=(num_gpu_blocks, num_cpu_blocks))
 
-        self.structured_output_manager = StructuredOutputManager(vllm_config)
+        if vllm_config.model_config.skip_tokenizer_init:
```
From the `StructuredOutputManager` init I see you initialize the very same tokenizer that is shipped with the model.

https://github.com/vllm-project/vllm/blob/022c58b80f8ae2e27dc526860b769455ef4c5498/vllm/v1/structured_output/__init__.py#L49-L53

which is exactly the same that happens in the init of the async engine.

In this case we are skipping the initialization of the tokenizer because the model is not shipped with a tokenizer in the first plance, and therefore the initialization of the `StructuredOutputManager` fails.  

Does this make sense? What would be the preferred way of handling this situation?

maxdebayser commented on vllm/model_executor/models/prithvi_geospatial_mae.py:
```diff
@@ -173,6 +188,13 @@ def _parse_and_validate_multimodal_data(
+        # the input prompt we pass one token and the size of the dummy
+        #  embedding tensors must reflect that.
+        return torch.empty(input_ids.shape)
+
```
I don't see the `get_multimodal_embeddings()` function here. How is this not failing in `GPUModelRunner.profile_run()`?

maxdebayser replied:
Yes, but if we're adding a new programming model for MM inputs I think it would be nice to handle this. If not a full solution, perhaps a test to disable cuda graphs and memory profiling for this special case.

But I'm thinking that maybe the raw MM inputs will be hard to generalize. Perhaps a hacky solution for this particular model would be to store the raw MM inputs in the `get_multimodal_embeddings()` method and then use them when `forward()` is called.

aarnphm reviewed (commented):
Ok, I did another round of review for other code misc to reduce general diff. I don't have an opinion for the KVCache part, so will leave that to others to comment on it.

heheda12345 commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -64,7 +65,131 @@ def new_empty(self) -> "KVCacheBlocks":
+
+    def get_computed_blocks(self,
+                            request: Request) -> tuple[KVCacheBlocks, int]:
+        return (KVCacheBlocks([]), 0)
```
What's your expect behavior of prefix caching? Seems that you want prefix caching to be enabled in "last" pooling method.
https://github.com/vllm-project/vllm/blob/8aeaa910a2c514022ab7521ce74eec0734886a1b/vllm/config.py#L4594

heheda12345 reviewed (commented):
Is the KVCacheManager modification necessary to this PR? As https://github.com/vllm-project/vllm/pull/16188 is runnable, is it possible to only include other necessary changes in this PR and split the KVCacheManager changes to another PR? It can both unblock this PR and help me to review KVCacheManager changes easier.

christian-pinto commented:
> Is the KVCacheManager modification necessary to this PR? As #16188 is runnable, is it possible to only include other necessary changes in this PR and split the KVCacheManager changes to another PR? It can both unblock this PR and help me to review KVCacheManager changes easier.

Hi @heheda12345, unfortunately these changes are necessary or otherwise I will have to hack around a number of things including the profiling run for memory profiling as this is not a language model so profiling as it is does not work. These changes would be totally useless because I do not need any of this profiling to take place as this model does not use attention. Therefore I would endup making some changes that would be there solely to support my model until attention free models are supported and then revert everything back as it was.
Does this make sense?

To help you in reviewing the two options, in the new PR (the one using zero blocks) the changes for supporting attention free models are only in the last commit. Similarly in this PR all changes related to supporting attention free are in commits 953f66ae3ffba7dd4e99b9ab286877a5d161e434 and 953f66ae3ffba7dd4e99b9ab286877a5d161e434

christian-pinto committed:
Some format chawnges to improve readability and diff sizes

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>
